### 다중선형회귀
#### 시간대비 성적데이터
+ 좀 더 정확한 예측을 위해서는 추가정보 필요
+ 공부시간외에 성적에 영향을 끼칠수 있는 다른 요인필요
+ 과외시간을 추가함 => 항이 추가 => 기울기 하나 더 추가
+ $ y = ax1 + bx2 + c $

In [3]:
import tensorflow as tf
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Dense
from tensorflow.keras.optimizers import RMSprop

In [2]:
times = [2,4,6,8]
lession= [0,4,2,3]
sungjuk = [81,93,91,97]

In [3]:
data = [[2,0,81],[4,4,93],[6,2,91],[8,3,97]]
xdata1 = [x[0] for x in data]
xdata2 = [x[1] for x in data]
ydata = [x[2] for x in data]

In [4]:
a = tf.Variable(tf.random_uniform([1], 0, 10,
                dtype=tf.float64, seed=1))
b = tf.Variable(tf.random_uniform([1], 0, 10,
                dtype=tf.float64, seed=1))
c = tf.Variable(tf.random_uniform([1], 0, 100,
                dtype=tf.float64, seed=1))

In [6]:
y = a * xdata1 + b * xdata2 + c

In [7]:
rmse = tf.sqrt(tf.reduce_mean(tf.square(y - ydata)))

In [8]:
learn_rate = 0.1

In [9]:
gradient_decent = tf.train.GradientDescentOptimizer\
                (learn_rate).minimize(rmse)

In [10]:
with tf.Session() as sess:
    sess.run(tf.global_variables_initializer())
    for step in range(2001):
        sess.run(gradient_decent)
        if step % 10 == 0:
            print('훈련횟수 epoch %.f, RMSE %.3f,'\
            '기울기1 %.3f, 기울기2 %.3f, 절편 %.3f' %
                  ( step, sess.run(rmse),
                    sess.run(a), sess.run(b),
                    sess.run(c) ))

2021-12-03 09:51:52.069576: I tensorflow/core/platform/cpu_feature_guard.cc:142] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA


훈련횟수 epoch 0, RMSE 67.394,기울기1 8.582, 기울기2 8.871, 절편 91.136
훈련횟수 epoch 10, RMSE 31.201,기울기1 3.204, 기울기2 6.344, 절편 90.187
훈련횟수 epoch 20, RMSE 5.508,기울기1 -1.120, 기울기2 4.052, 절편 89.288
훈련횟수 epoch 30, RMSE 5.025,기울기1 -0.919, 기울기2 3.503, 절편 88.941
훈련횟수 epoch 40, RMSE 4.666,기울기1 -0.670, 기울기2 3.084, 절편 88.590
훈련횟수 epoch 50, RMSE 4.378,기울기1 -0.458, 기울기2 2.753, 절편 88.225
훈련횟수 epoch 60, RMSE 4.142,기울기1 -0.280, 기울기2 2.500, 절편 87.850
훈련횟수 epoch 70, RMSE 3.939,기울기1 -0.131, 기울기2 2.313, 절편 87.468
훈련횟수 epoch 80, RMSE 3.757,기울기1 -0.006, 기울기2 2.179, 절편 87.082
훈련횟수 epoch 90, RMSE 3.586,기울기1 0.102, 기울기2 2.087, 절편 86.693
훈련횟수 epoch 100, RMSE 3.421,기울기1 0.195, 기울기2 2.027, 절편 86.303
훈련횟수 epoch 110, RMSE 3.261,기울기1 0.277, 기울기2 1.991, 절편 85.912
훈련횟수 epoch 120, RMSE 3.101,기울기1 0.352, 기울기2 1.972, 절편 85.520
훈련횟수 epoch 130, RMSE 2.943,기울기1 0.420, 기울기2 1.965, 절편 85.129
훈련횟수 epoch 140, RMSE 2.786,기울기1 0.485, 기울기2 1.968, 절편 84.737
훈련횟수 epoch 150, RMSE 2.629,기울기1 0.547, 기울기2 1.976, 절편 84.346
훈련횟수 epoch 160, RMSE 2.472

In [ ]:
# -> epoch 1030부터 기울기와 절편이 일정하게 출력
# y = 1.77x + 2.4x2 + 77.9

In [16]:
learn_rate = 0.5

In [17]:
gradient_decent = tf.train.GradientDescentOptimizer\
                (learn_rate).minimize(rmse)

In [18]:
with tf.Session() as sess:
    sess.run(tf.global_variables_initializer())
    for step in range(2001):
        sess.run(gradient_decent)
        if step % 10 == 0:
            print('훈련횟수 epoch %.f, RMSE %.3f,'\
            '기울기1 %.3f, 기울기2 %.3f, 절편 %.3f' %
                  ( step, sess.run(rmse),
                    sess.run(a), sess.run(b),
                    sess.run(c) ))

훈련횟수 epoch 0, RMSE 52.814,기울기1 6.417, 기울기2 7.865, 절편 90.759
훈련횟수 epoch 10, RMSE 9.166,기울기1 0.535, 기울기2 3.561, 절편 88.760
훈련횟수 epoch 20, RMSE 9.173,기울기1 1.022, 기울기2 2.942, 절편 87.886
훈련횟수 epoch 30, RMSE 9.177,기울기1 1.312, 기울기2 2.677, 절편 87.077
훈련횟수 epoch 40, RMSE 9.179,기울기1 1.507, 기울기2 2.575, 절편 86.334
훈련횟수 epoch 50, RMSE 9.180,기울기1 1.654, 기울기2 2.548, 절편 85.650
훈련횟수 epoch 60, RMSE 9.181,기울기1 1.773, 기울기2 2.553, 절편 85.023
훈련횟수 epoch 70, RMSE 9.182,기울기1 1.875, 기울기2 2.572, 절편 84.448
훈련횟수 epoch 80, RMSE 9.182,기울기1 1.964, 기울기2 2.596, 절편 83.920
훈련횟수 epoch 90, RMSE 9.183,기울기1 2.044, 기울기2 2.620, 절편 83.436
훈련횟수 epoch 100, RMSE 9.183,기울기1 2.115, 기울기2 2.644, 절편 82.992
훈련횟수 epoch 110, RMSE 9.183,기울기1 2.180, 기울기2 2.666, 절편 82.585
훈련횟수 epoch 120, RMSE 9.184,기울기1 2.239, 기울기2 2.687, 절편 82.212
훈련횟수 epoch 130, RMSE 9.184,기울기1 2.293, 기울기2 2.705, 절편 81.869
훈련횟수 epoch 140, RMSE 9.184,기울기1 2.342, 기울기2 2.722, 절편 81.555
훈련횟수 epoch 150, RMSE 9.184,기울기1 2.386, 기울기2 2.738, 절편 81.267
훈련횟수 epoch 160, RMSE 9.184,기울기1 2.

In [19]:
learn_rate = 0.05

In [20]:
gradient_decent = tf.train.GradientDescentOptimizer\
                (learn_rate).minimize(rmse)

In [21]:
with tf.Session() as sess:
    sess.run(tf.global_variables_initializer())
    for step in range(2001):
        sess.run(gradient_decent)
        if step % 10 == 0:
            print('훈련횟수 epoch %.f, RMSE %.3f,'\
            '기울기1 %.3f, 기울기2 %.3f, 절편 %.3f' %
                  ( step, sess.run(rmse),
                    sess.run(a), sess.run(b),
                    sess.run(c) ))

훈련횟수 epoch 0, RMSE 69.218,기울기1 8.852, 기울기2 8.997, 절편 91.183
훈련횟수 epoch 10, RMSE 51.017,기울기1 6.152, 기울기2 7.737, 절편 90.711
훈련횟수 epoch 20, RMSE 32.993,기울기1 3.472, 기울기2 6.470, 절편 90.234
훈련횟수 epoch 30, RMSE 15.722,기울기1 0.863, 기울기2 5.200, 절편 89.751
훈련횟수 epoch 40, RMSE 5.663,기울기1 -1.012, 기울기2 4.124, 절편 89.318
훈련횟수 epoch 50, RMSE 5.255,기울기1 -1.066, 기울기2 3.767, 절편 89.122
훈련횟수 epoch 60, RMSE 5.040,기울기1 -0.930, 기울기2 3.521, 절편 88.953
훈련횟수 epoch 70, RMSE 4.849,기울기1 -0.800, 기울기2 3.300, 절편 88.779
훈련횟수 epoch 80, RMSE 4.678,기울기1 -0.680, 기울기2 3.101, 절편 88.602
훈련횟수 epoch 90, RMSE 4.526,기울기1 -0.569, 기울기2 2.924, 절편 88.421
훈련횟수 epoch 100, RMSE 4.389,기울기1 -0.467, 기울기2 2.768, 절편 88.237
훈련횟수 epoch 110, RMSE 4.265,기울기1 -0.374, 기울기2 2.631, 절편 88.050
훈련횟수 epoch 120, RMSE 4.151,기울기1 -0.288, 기울기2 2.512, 절편 87.862
훈련횟수 epoch 130, RMSE 4.045,기울기1 -0.210, 기울기2 2.410, 절편 87.671
훈련횟수 epoch 140, RMSE 3.947,기울기1 -0.138, 기울기2 2.323, 절편 87.480
훈련횟수 epoch 150, RMSE 3.853,기울기1 -0.072, 기울기2 2.249, 절편 87.287
훈련횟수 epoch 160, RMS

#### 0.1 - 훈련횟수 epoch 2000, RMSE 1.837,기울기1 1.770, 기울기2 2.408, 절편 77.903
#### 0.5 - 훈련횟수 epoch 2000, RMSE 9.185,기울기1 2.863, 기울기2 2.904, 절편 78.087
#### 0.05 - 훈련횟수 epoch 2000, RMSE 0.919,기울기1 1.630, 기울기2 2.345, 절편 77.879
#### 모델 분석 수식과 하이퍼 파라미터 설정이 중요
+ rmse 수식

In [41]:
import numpy as np

In [48]:
xdata = np.array([x[:1] for x in data])
ydata = np.array([x[2] for x in data])

In [49]:
model = Sequential()
model.add(Dense(1, input_dim=1))

In [50]:
model.compile(loss='mse', optimizer='sgd',
              metrics=['accuracy'])

In [55]:
model.fit(xdata, ydata, epochs=1000, verbose=1)

Train on 4 samples
Epoch 1/1000
4/4 [==============================] - 0s 320us/sample - loss: 8.3022 - acc: 0.0000e+00
Epoch 2/1000
4/4 [==============================] - 0s 3ms/sample - loss: 8.3022 - acc: 0.0000e+00
Epoch 3/1000
4/4 [==============================] - 0s 269us/sample - loss: 8.3022 - acc: 0.0000e+00
Epoch 4/1000
4/4 [==============================] - 0s 268us/sample - loss: 8.3022 - acc: 0.0000e+00
Epoch 5/1000
4/4 [==============================] - 0s 257us/sample - loss: 8.3022 - acc: 0.0000e+00
Epoch 6/1000
4/4 [==============================] - 0s 256us/sample - loss: 8.3022 - acc: 0.0000e+00
Epoch 7/1000
4/4 [==============================] - 0s 259us/sample - loss: 8.3021 - acc: 0.0000e+00
Epoch 8/1000
4/4 [==============================] - 0s 258us/sample - loss: 8.3021 - acc: 0.0000e+00
Epoch 9/1000
4/4 [==============================] - 0s 333us/sample - loss: 8.3021 - acc: 0.0000e+00
Epoch 10/1000
4/4 [==============================] - 0s 310us/sample - los

In [56]:
print('기울기 %.3f' % model.get_weights()[0])
print('절편 %.3f' % model.get_weights()[1])

기울기 2.301
절편 78.995


### 기울기

In [60]:
model.get_weights()[0][0], model.get_weights()[0][1]

IndexError: index 1 is out of bounds for axis 0 with size 1

In [59]:
model.get_weights()[1]

array([78.9954], dtype=float32)

In [61]:
model.fit(xdata, ydata, epochs=2000, verbose=0)

In [62]:
print('기울기 %.3f' % model.get_weights()[0])
print('절편 %.3f' % model.get_weights()[1])

기울기 2.300
절편 78.999


In [63]:
model.get_weights()[0][0], model.get_weights()[0][1]

IndexError: index 1 is out of bounds for axis 0 with size 1

#### 학습된 모델 저장과 재사용
#### 학습이 끝난후 테스트 결과가 만족스러울때 이를 파일(모델)로 저장해두면 새로운 데이터에 바로 사용 가능

In [65]:
pip install np_utils

Defaulting to user installation because normal site-packages is not writeable
     |████████████████████████████████| 61 kB 1.6 MB/s eta 0:00:01
     |████████████████████████████████| 829 kB 3.0 MB/s eta 0:00:01
  Created wheel for np-utils: filename=np_utils-0.5.12.1-py3-none-any.whl size=57131 sha256=2803771420babd0dedfe33c730437a45b34d6849944bca6e6af7012bff385e8f
  Stored in directory: /home/hadoop/.cache/pip/wheels/8c/4e/ef/095c24693723c329f4cdc1079861cdbb2487d4b41b2496a4e7
  Created wheel for future: filename=future-0.18.2-py3-none-any.whl size=491070 sha256=c668bcfc10ebe50c100ec93b9f078896c77871a84f07ec3c93ea4843f378fd99
  Stored in directory: /home/hadoop/.cache/pip/wheels/56/b0/fe/4410d17b32f1f0c3cf54cdfb2bc04d7b4b8f4ae377e2229ba0
Successfully built np-utils future
  Consider adding this directory to PATH or, if you prefer to suppress this warning, use --no-warn-script-location.
Note: you may need to restart the kernel to use updated packages.


In [2]:
from tensorflow.keras import utils
from sklearn.preprocessing import LabelEncoder
from tensorflow.keras.models import load_model
import pandas as pd

In [68]:
iris = pd.read_csv('data/iris.csv')

In [71]:
iris = pd.read_csv('data/iris.csv', header=None)

x_data = iris.iloc[:, 0:4]
y_data = iris.iloc[:, 4]


le = LabelEncoder()
le.fit(y_data)
y_data = le.transform(y_data)
y_data = utils.to_categorical(y_data)

#### 시드 설정

In [72]:
seed = 211202
np.random.seed(seed)
tf.set_random_seed(seed)

#### 신경망, 손실함수, 옵티마이저 설정

In [73]:
model = Sequential()
model.add(Dense(10, input_dim=4, activation='relu'))
model.add(Dense(3, activation='softmax'))

model.compile(loss='categorical_crossentropy',
              optimizer='adam',
              metrics=['accuracy', 'mse'])

In [74]:
model.fit(x_data, y_data, epochs=1000, verbose=0)

In [75]:
print('%.4f' % model.evaluate(x_data, y_data)[1])

150/150 [==============================] - 0s 306us/sample - loss: 0.0762 - acc: 0.9800 - mean_squared_error: 0.0126
0.9800


#### 모델 저장 : 객체.save(경로)

In [77]:
model.save('models/iris.h5') # 모델 확장자 h5

#### 모델 불러오기 : load_model(경로)

In [6]:
model = load_model('models/iris.h5')

Instructions for updating:
Call initializer instance with the dtype argument instead of passing it to the constructor
Instructions for updating:
Call initializer instance with the dtype argument instead of passing it to the constructor
Instructions for updating:
If using Keras pass *_constraint arguments to layers.


2021-12-03 11:02:31.491611: I tensorflow/core/platform/cpu_feature_guard.cc:142] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 FMA


In [7]:
print('%.4f' % model.evaluate(x_data, y_data)[1])

new_x = np.array([[6.4, 2.7, 5.3, 1.9]])
y_predict = model.predict(new_x)
print(y_predict[0][0],y_predict[0][1],y_predict[0][2])

NameError: name 'x_data' is not defined

### 과적합 문제 해결
+ 1998년 존스홉킨스 대학 세즈노프스키 교수가 힌튼 교수가 발표한 역전파 알고리즘에 과심을 가짐

+ 즉, 은닉층과 역전파의 효과가 얼마나 큰지 궁금

+ 따라서, 광석과 일반돌에 음파를 쏘아 측정치 조사

+ 과연 이 돌을 잘 분류할 수 있을지 실험(sonar.csv)

#### 논문결과 : 은닉층의 수가 올라갈수록 데이터 예측 정확도도 같이 상승
+  학습/검증 데이터로 분리해서 진행하면 증가속도가 다소 지연됨

#### 과적합 방지 방법 - 데이터를 train/test로 분리해서 훈련 실시 (+ 교차검증)

In [9]:
sonar = pd.read_csv('data/sonar.csv', header=None)

In [5]:
import numpy as np

In [10]:
x_data = sonar.iloc[:, 0:60]
y_data = sonar.iloc[:, 60]
print(y_data[:5])

0    R
1    R
2    R
3    R
4    R
Name: 60, dtype: object


In [11]:
sonar.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 208 entries, 0 to 207
Data columns (total 61 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   0       208 non-null    float64
 1   1       208 non-null    float64
 2   2       208 non-null    float64
 3   3       208 non-null    float64
 4   4       208 non-null    float64
 5   5       208 non-null    float64
 6   6       208 non-null    float64
 7   7       208 non-null    float64
 8   8       208 non-null    float64
 9   9       208 non-null    float64
 10  10      208 non-null    float64
 11  11      208 non-null    float64
 12  12      208 non-null    float64
 13  13      208 non-null    float64
 14  14      208 non-null    float64
 15  15      208 non-null    float64
 16  16      208 non-null    float64
 17  17      208 non-null    float64
 18  18      208 non-null    float64
 19  19      208 non-null    float64
 20  20      208 non-null    float64
 21  21      208 non-null    float64
 22  22

In [12]:
print(y_data[:5]) # R(rock), M(mine)

0    R
1    R
2    R
3    R
4    R
Name: 60, dtype: object


In [14]:
sonar.tail()

0       1       2       3       4       5       6       7       8   \
203  0.0187  0.0346  0.0168  0.0177  0.0393  0.1630  0.2028  0.1694  0.2328   
204  0.0323  0.0101  0.0298  0.0564  0.0760  0.0958  0.0990  0.1018  0.1030   
205  0.0522  0.0437  0.0180  0.0292  0.0351  0.1171  0.1257  0.1178  0.1258   
206  0.0303  0.0353  0.0490  0.0608  0.0167  0.1354  0.1465  0.1123  0.1945   
207  0.0260  0.0363  0.0136  0.0272  0.0214  0.0338  0.0655  0.1400  0.1843   

         9   ...      51      52      53      54      55      56      57  \
203  0.2684  ...  0.0116  0.0098  0.0199  0.0033  0.0101  0.0065  0.0115   
204  0.2154  ...  0.0061  0.0093  0.0135  0.0063  0.0063  0.0034  0.0032   
205  0.2529  ...  0.0160  0.0029  0.0051  0.0062  0.0089  0.0140  0.0138   
206  0.2354  ...  0.0086  0.0046  0.0126  0.0036  0.0035  0.0034  0.0079   
207  0.2354  ...  0.0146  0.0129  0.0047  0.0039  0.0061  0.0040  0.0036   

         58      59  60  
203  0.0193  0.0157   M  
204  0.0062  0.0067   M  
205  0.0077  0.0031   M  
206  0.0036  0.0048   M  
207  0.0061  0.0115   M  

[5 rows x 61 columns]

In [45]:
# 전처리 수행 - 범주형 데이터 변환

le = LabelEncoder()
y_data = le.fit_transform(y_data)
print(y_data[:5])

[1 1 1 1 1]


In [46]:
# 난수 지정
seed = 190121
np.random.seed(seed)
tf.set_random_seed(seed)

In [47]:
# 신경망 구성 : 
# 특성수 : 60
# 입력노드 : ???, 활성화 함수 : relu
# 최종 출력노드 : 1, 활성화 함수 : sigmoid
model = Sequential()

#### 은닉층 설정 예시

In [21]:
model.add(Dense(40, input_dim=60, activation='relu')) # 입력노드 60개를 40으로 줄임

In [ ]:
model.add(Dense(20, input_dim=40, activation='relu')) # 은닉층 : 40개를 20으로 줄임

In [51]:
model.add(Dense(1, activation='sigmoid')) # 최종 노드 1개

In [48]:
model.add(Dense(25, input_dim=60, activation='relu'))

In [ ]:
model.add(Dense(1, activation='sigmoid'))

In [49]:
# 손실함수, 최적화함수 정의
model.compile(loss='binary_crossentropy',  # 이항분류 - 둘 중 하나 분류
              optimizer='adam', # RMSPROP + MOMENTUM
              metrics=['accuracy','mse']) # 가중치 설정 확인용 정확도와 오차

In [52]:
model.fit(x_data, y_data, epochs=1000, verbose=0)

In [53]:
print(model.evaluate(x_data, y_data)[1]) # 결과 1.0 과적합 발생 가능성 높음

208/208 [==============================] - 0s 223us/sample - loss: 0.0194 - acc: 1.0000 - mean_squared_error: 0.0017
1.0


In [30]:
from sklearn.model_selection import train_test_split

In [54]:
# 과적합을 방지하기 위해
# 데이터셋에서 train/test 데이터로 분리
xtrain, xtest, ytrain, ytest = \
     train_test_split(x_data, y_data,
                      test_size=0.3,
                      random_state=seed)

model.fit(xtrain, ytrain, epochs=1000, verbose=0)

print(model.evaluate(xtest, ytest)[1])

63/63 [==============================] - 0s 62us/sample - loss: 0.0876 - acc: 0.9524 - mean_squared_error: 0.0293
0.95238096


In [55]:
# 모델저장
model.save('models/sonar.csv')